From a8b6fc647f41543f83ba24867d5e9142bc06f880 Mon Sep 17 00:00:00 2001
From: Jingcheng Du <jingcheng.du@intel.com>
Date: Fri, 24 Jun 2016 14:39:53 +0800
Subject: [PATCH 318/362] HBASE-15353 Add metric for number of
 CallQueueTooBigException's

(cherry picked from commit 687c70c96339fccace6db1e0f327d11fe9f0c2a4)

Change-Id: I87161e5580a3ff455706e43c5aa570e67c631579
Author: Jingcheng Du
Reason: Improvement
Ref: CDH-51762
---
 .../hadoop/hbase/ipc/MetricsHBaseServerSource.java |    3 +++
 .../hbase/ipc/MetricsHBaseServerSourceImpl.java    |    8 ++++++++
 .../hadoop/hbase/ipc/MetricsHBaseServer.java       |    3 +++
 3 files changed, 14 insertions(+)

diff --git a/hbase-hadoop-compat/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServerSource.java b/hbase-hadoop-compat/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServerSource.java
index 80ce984..a3ffcb2 100644
--- a/hbase-hadoop-compat/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServerSource.java
+++ b/hbase-hadoop-compat/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServerSource.java
@@ -78,6 +78,8 @@ public interface MetricsHBaseServerSource extends BaseSource {
   String EXCEPTIONS_MULTI_TOO_LARGE_NAME = "exceptions.multiResponseTooLarge";
   String EXCEPTIONS_MULTI_TOO_LARGE_DESC = "A response to a multi request was too large and the " +
       "rest of the requests will have to be retried.";
+  String EXCEPTIONS_CALL_QUEUE_TOO_BIG = "exceptions.callQueueTooBig";
+  String EXCEPTIONS_CALL_QUEUE_TOO_BIG_DESC = "Call queue is full";
 
   void authorizationSuccess();
 
@@ -102,6 +104,7 @@ public interface MetricsHBaseServerSource extends BaseSource {
   void scannerResetException();
   void tooBusyException();
   void multiActionTooLargeException();
+  void callQueueTooBigException();
 
   void sentBytes(long count);
 
diff --git a/hbase-hadoop2-compat/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServerSourceImpl.java b/hbase-hadoop2-compat/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServerSourceImpl.java
index 2a25e20..857ffbe 100644
--- a/hbase-hadoop2-compat/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServerSourceImpl.java
+++ b/hbase-hadoop2-compat/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServerSourceImpl.java
@@ -50,6 +50,7 @@ public class MetricsHBaseServerSourceImpl extends BaseSourceImpl
   private final MutableFastCounter exceptionsNSRE;
   private final MutableFastCounter exceptionsMoved;
   private final MutableFastCounter exceptionsMultiTooLarge;
+  private final MutableFastCounter exceptionsCallQueueTooBig;
 
 
   private MetricHistogram queueCallTime;
@@ -88,6 +89,8 @@ public class MetricsHBaseServerSourceImpl extends BaseSourceImpl
         .newCounter(EXCEPTIONS_NSRE_NAME, EXCEPTIONS_TYPE_DESC, 0L);
     this.exceptionsMultiTooLarge = this.getMetricsRegistry()
         .newCounter(EXCEPTIONS_MULTI_TOO_LARGE_NAME, EXCEPTIONS_MULTI_TOO_LARGE_DESC, 0L);
+    this.exceptionsCallQueueTooBig = this.getMetricsRegistry().newCounter(
+      EXCEPTIONS_CALL_QUEUE_TOO_BIG, EXCEPTIONS_CALL_QUEUE_TOO_BIG_DESC, 0L);
 
     this.authenticationSuccesses = this.getMetricsRegistry().newCounter(
         AUTHENTICATION_SUCCESSES_NAME, AUTHENTICATION_SUCCESSES_DESC, 0L);
@@ -177,6 +180,11 @@ public class MetricsHBaseServerSourceImpl extends BaseSourceImpl
   }
 
   @Override
+  public void callQueueTooBigException() {
+    exceptionsCallQueueTooBig.incr();
+  }
+
+  @Override
   public void authenticationSuccess() {
     authenticationSuccesses.incr();
   }
diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServer.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServer.java
index 47642a5..fe03d4f 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServer.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/MetricsHBaseServer.java
@@ -19,6 +19,7 @@
 
 package org.apache.hadoop.hbase.ipc;
 
+import org.apache.hadoop.hbase.CallQueueTooBigException;
 import org.apache.hadoop.hbase.MultiActionResultTooLarge;
 import org.apache.hadoop.hbase.NotServingRegionException;
 import org.apache.hadoop.hbase.RegionTooBusyException;
@@ -113,6 +114,8 @@ public class MetricsHBaseServer {
         source.failedSanityException();
       } else if (throwable instanceof MultiActionResultTooLarge) {
         source.multiActionTooLargeException();
+      } else if (throwable instanceof CallQueueTooBigException) {
+        source.callQueueTooBigException();
       }
     }
   }
-- 
1.7.9.5

